libraries/ms-common/src/stm32f0xx/interrupt.c

```diff
@@ -0,0 +1,200 @@
+#include "interrupt.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "gpio.h"
+#include "gpio_cfg.h"
+#include "status.h"
+#include "stm32f0xx_exti.h"
+#include "stm32f0xx_misc.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_syscfg.h"
+
+#define NUM_INTERRUPT_GPIO_CHANNELS 3
+static InterruptPriority s_interrupt_gpio_priority_map[NUM_INTERRUPT_GPIO_CHANNELS];
+
+void interrupt_init() {
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_SYSCFG, ENABLE);
+
+  for (int i = 0; i < NUM_INTERRUPT_GPIO_CHANNELS; i++) {
+    s_interrupt_gpio_priority_map[i] = NUM_INTERRUPT_PRIORITY;
+  }
+}
+
+#pragma GCC push_options
+#pragma GCC optimize("O0")
+// Use this to count how many nested disables there are and only enable once out of all critical
+// sections.
+static uint8_t s_interrupt_priority_mask = 0;
+
+void interrupt_enable() {
+  if (s_interrupt_priority_mask > 0) {
+    s_interrupt_priority_mask--;
+  }
+  if (!s_interrupt_priority_mask) {
+    __enable_irq();
+  }
+}
+
+void interrupt_disable() {
+  s_interrupt_priority_mask++;
+  __disable_irq();
+}
+#pragma GCC pop_options
+
+// REGISTRATION AND TRIGGER HELPERS
+// Use these functions to register and trigger interrupts using module specific registration and
+// trigger functions.
+
+#define NUM_INTERRUPT_CHANNELS 32
+
+// Enables the nested interrupt vector controller for a given EXTI line
+StatusCode prv_nvic_enable(uint8_t irq_channel, InterruptPriority priority) {
+  if (priority >= NUM_INTERRUPT_PRIORITY || irq_channel >= NUM_INTERRUPT_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  NVIC_InitTypeDef init_struct = { .NVIC_IRQChannel = irq_channel,
+                                  .NVIC_IRQChannelPriority = priority,
+                                  .NVIC_IRQChannelCmd = ENABLE };
+
+  NVIC_Init(&init_struct);
+
+  return STATUS_CODE_OK;
+}
+
+#define NUM_INTERRUPT_LINES 32
+
+// Enables the interrupt line with the given settings.
+StatusCode prv_exti_enable(uint8_t line, InterruptSettings *settings) {
+  if (line >= NUM_INTERRUPT_LINES || settings->type >= NUM_INTERRUPT_TYPE ||
+      settings->edge >= NUM_INTERRUPT_EDGE) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  EXTI_InitTypeDef init_struct = { .EXTI_Line = 0x01 << line,
```

Comment:
Ah, it's just a case of personal preference again. I tend to format struct declarations like
````
type foo = {
  .bar = val,
  .other = thing
};
````
to mirror struct declarations.